src/mbgl/style/expression/distance.cpp

```diff
@@ -0,0 +1,601 @@
+    if (!params.feature || !params.canonical) {
+        return EvaluationError{"distance expression requirs valid feature and canonical information."};
+    }
+    auto geometryType = params.feature->getType();
+    if (geometryType == FeatureType::Point || geometryType == FeatureType::LineString) {
+        auto distance = calculateDistance(*params.feature, *params.canonical, geometries, unit);
+        if (!std::isnan(distance)) {
+            return distance;
+        }
+    }
+    return EvaluationError{"distance expression currently only evaluates Point/LineString geometries."};
+}
+
+ParseResult Distance::parse(const Convertible& value, ParsingContext& ctx) {
+    auto parsedValue = parseValue(value, ctx);
+    if (!parsedValue) {
+        return ParseResult();
+    }
+
+    return parsedValue->geojson.match(
+        [&parsedValue, &ctx](const mapbox::geometry::geometry<double>& geometrySet) {
+            if (auto ret = getGeometry(mbgl::Feature(geometrySet), ctx)) {
+                return ParseResult(
+                    std::make_unique<Distance>(parsedValue->geojson, std::move(*ret), parsedValue->unit));
+            }
+            return ParseResult();
+        },
+        [&parsedValue, &ctx](const mapbox::feature::feature<double>& feature) {
+            if (auto ret = getGeometry(mbgl::Feature(feature), ctx)) {
+                return ParseResult(
+                    std::make_unique<Distance>(parsedValue->geojson, std::move(*ret), parsedValue->unit));
+            }
+            return ParseResult();
+        },
+        [&parsedValue, &ctx](const mapbox::feature::feature_collection<double>& features) {
+            for (const auto& feature : features) {
+                if (auto ret = getGeometry(mbgl::Feature(feature), ctx)) {
+                    return ParseResult(
+                        std::make_unique<Distance>(parsedValue->geojson, std::move(*ret), parsedValue->unit));
+                }
+            }
+            return ParseResult();
+        },
+        [&ctx](const auto&) {
+            ctx.error("'distance' expression requires valid geojson that contains LineString/Point geometries.");
+            return ParseResult();
+        });
+
+    return ParseResult();
+}
+
+Value convertValue(const mapbox::geojson::rapidjson_value& v) {
+    if (v.IsDouble()) {
+        return v.GetDouble();
+    }
+    if (v.IsString()) {
+        return std::string(v.GetString());
+    }
+    if (v.IsArray()) {
+        std::vector<Value> result;
+        result.reserve(v.Size());
+        for (const auto& m : v.GetArray()) {
+            result.push_back(convertValue(m));
+        }
+        return result;
+    }
+    if (v.IsObject()) {
+        std::unordered_map<std::string, Value> result;
+        for (const auto& m : v.GetObject()) {
+            result.emplace(m.name.GetString(), convertValue(m.value));
+        }
+        return result;
+    }
+    // Ignore other types as valid geojson only contains above types.
```

Comment:
geojson can contain features object that is an any [JSON or Null](https://tools.ietf.org/html/rfc7946#section-3.2). Someone might use features to keep additional information.

Reply:
As it is a recursive call, in case it is `null` value, the object will have `Null` value inserted, so I assume it won't be a problem.

But I tried to test will geojson that has `null` geometry object, it turns out our geojson_impl.hpp currently does not support geometry with null value.

Reply:
@zmiao sorry, I didn't explain clearly what I meant. If developer sets `['distance', GEOJson]` where `GEOJson` is a valid geojson with feature properties, do we expect to get it back when we do serialize? In other words would expression round-trip precisely?

Pseudocode: `['distance', GEOJson]` == `createExpression(['distance', GEOJson]->serialize())`

Reply:
Yes, the purpose is to serialize all valid contents, i.e. keep every member of the objects, but the member's order may change. The round-trip is verified by expression tests:  mapbox/mapbox-gl-js#9602

Reply:
@zmiao could you add feature properties to https://github.com/mapbox/mapbox-gl-js/pull/9602/files#diff-b53dbf3009e15d7acfb1f8871bf7028dR3 ?

Reply:
I see the problem. Thank you so much! Now it is fixed.  Tests updated accordingly: https://github.com/mapbox/mapbox-gl-js/pull/9602/files#diff-b53dbf3009e15d7acfb1f8871bf7028dR3